### PR TITLE
[SYCL][FPGA] Refactor of [[intel::bankwidth()]] and [[intel::numbanks()]] attributes

### PR DESCRIPTION
**Pull request by smanna12**
This patch

1. refactors two FPGA memory attributes to align with upstream:
    [[intel::bankwidth()]]
    [[intel::numbanks()]]
 
2. refactors the way we handle duplicate attributes and
    mutually exclusive attributes logic when present on a given declaration.

3. handles redeclarations or template instantiations properly.
4. adds test
5. removes generic function "AddOneConstantPowerTwoValueAttr"

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -6409,21 +6409,178 @@ static void handleIntelFPGARegisterAttr(Sema &S, Decl *D, const ParsedAttr &A) {
...
+      unsigned NumBankBits = BBA->args_size();
+      if (NumBankBits != ArgVal.ceilLogBase2()) {
+        Diag(E->getExprLoc(), diag::err_bankbits_numbanks_conflicting) << CI;
+        return;
+      }
+    }
+
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<IntelFPGANumBanksAttr>()) {
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:6532]: Do we have to make sure that we're not checking the implicitly added num banks attribute that we drop a few lines below? (We may need additional test coverage to prove that.)

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -6389,10 +6389,6 @@ static bool checkIntelFPGARegisterAttrCompatibility(Sema &S, Decl *D,
     if (!MA->isImplicit() &&
         checkAttrMutualExclusion<IntelFPGAMemoryAttr>(S, D, Attr))
       InCompat = true;
-  if (auto *NBA = D->getAttr<IntelFPGANumBanksAttr>())
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:6392]: This change has me a bit worried (along with the new `MutualExclusions` definition in Attr.td) -- it used to be we'd only look for a mutual exclusion when the `numbanks` attribute was not implicit. Now we're checking it always, but we still create a numbanks implicitly (see `Sema::AddIntelFPGABankBitsAttr()`).

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 3]
```diff
@@ -554,6 +587,17 @@ void diagnostics()
...
+  //CHECK: IntelFPGABankBitsAttr
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 2
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 2
+  //CHECK-NEXT: ConstantExpr{{.*}}'int'
+  //CHECK-NEXT: value: Int 3
+  //CHECK-NEXT: IntegerLiteral{{.*}}'int' 3
   //expected-error@+2 1{{'fpga_register' and 'bank_bits' attributes are not compatible}}
   [[intel::bank_bits(2, 3)]]
   [[intel::fpga_register]]
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-local.cpp:603]: @AaronBallman, thanks for the reviews. There is an existing test where we are not checking implicit Numbank Attribute during incompatible check with register and bank_bits attribute. I have added AST part to show that bank bits attribute generates  implicit Numbank but does not get checked. Is this sufficient for the test coverage?

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 2 above)
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:6392]: Sorry I just noticed your message. 
>>Now we're checking it always, but we still create a numbanks implicitly (see Sema::AddIntelFPGABankBitsAttr()).

Yes, correct

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:6392]: I checked AST for this test case without refactoring patch:
//expected-error@+2 1{{'fpga_register' and 'bank_bits' attributes are not compatible}}
  [[intel::bank_bits(2, 3)]]
  [[intel::fpga_register]]
  //expected-note@-2 1{{conflicting attribute is here}}
  unsigned int bb_reg[4];


```
 |-DeclStmt 0x5592c9ea9680 <line:558:3, line:561:25>
|   | `-VarDecl 0x5592c9ea94d0 <col:3, col:24> col:16 bb_reg 'unsigned int [4]'
|   |   |-IntelFPGANumBanksAttr 0x5592c9ea95b0 <<invalid sloc>> Implicit
|   |   | `-IntegerLiteral 0x5592c9ea9590 <<invalid sloc>> 'int' 4
|   |   |-IntelFPGAMemoryAttr 0x5592c9ea9610 <<invalid sloc>> Implicit memory Default
|   |   `-IntelFPGABankBitsAttr 0x5592c9ea9638 <line:558:5, col:26>
|   |     |-ConstantExpr 0x5592c9ea9550 <col:22> 'int'
|   |     | |-value: Int 2
|   |     | `-IntegerLiteral 0x5592c9ea9428 <col:22> 'int' 2
|   |     `-ConstantExpr 0x5592c9ea9570 <col:25> 'int'
|   |       |-value: Int 3
|   |       `-IntegerLiteral 0x5592c9ea9448 <col:25> 'int' 3

```

The same happens with current factoring patch.

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:6392]: Okay, thank you for checking!

**Review by AaronBallman — APPROVED** [review]
LGTM

**Pull request merged**